[claim.rb] feat: add new claim module and associated changes to xmlparser

FarFigNewGut opened the pull request:
the `Group` and `Clustered` callouts are kinda holdovers from O's original script not sure if that should just get taken out. but believe this is working. I'm still not running 5.7 as a daily driver but was fine in some impromptu testing and was a port of code i've been running in slightly older lich for months without issue. but there were some changes in xmlparser so might need extra verification.

This would expose `Claim.mine?` that could be use in bigshot et. all checks as well as letting other scripts block on `Claim.last_room == <target room>` to minimize race condition issues parsing room objects, etc.

mrhoribu commented on lib/claim.rb:
```diff
@@ -0,0 +1,93 @@
+      @mine = true
+      self.claim_room nav_rm unless nav_rm.nil?
+    rescue StandardError => e
+      Log.out(e)
```
This needs a `if defined? Log` added to the end of it otherwise will cause issues. Could do something like:
```ruby
if defined? Log
  Log.out(e)
else
  respond("Claim Parser Error: #{e}")
end

mrhoribu reviewed (changes requested):
Looks good. Just missed one reference to `Log`. I feel it's likely safe to leave in the Group and Cluster for now as we try to target migrating more of Olib over to core Lich5. Group being one of the ones I want to get added asap.

ondreian commented on lib/claim.rb:
```diff
@@ -0,0 +1,97 @@
+            'Checked' => Claim.checked?,
+            'Last Room' => Claim.last_room,
+            'Others' => Claim.others}
+    respond JSON.pretty_generate(info)
```
we can/should terminal table this output

mrhoribu replied:
This is used for primarily debugging/info purposes, so I'm ok with adding description column.

ondreian commented on lib/xmlparser.rb:
```diff
@@ -123,6 +125,11 @@ def initialize
     @previous_nav_rm = nil
+
+    # claim update
+    @arrival_pcs = []
```
This seems like something we should make an attr_reader so it's able to be read from outside this module probably?

FarFigNewGut replied:
attr_accessored arrival_pcs and the obvious_hidden_player values

FarFigNewGut commented:
- added the arrival_pcs and hidden_player values as attr_accessors in xmlparser
- tableized the Claim.info call (added a description field, might be too much now)
- also added exception handling for the Grouped and Clustered calls. I mean Log is still in there should probably chage that to Lich.log i guess but i figured Grouped especially might conflict with people's personal stuff so until we have an actual module to  CLAIM (GET IT?!?!) that namespace formally i figured i'd try to account for that.

mrhoribu commented:
The only other consideration I would have is should we name scope this to something other than just `Claim`? Should it be something like `Lich::Claim` incase of namespace conflicts with any other code that may use a module with similar name?

mrhoribu reviewed (approved):
lgtm. Eventually we'll look at adding callbacks, but for now, this works.
